fish-rust/src/curses.rs

```diff
@@ -0,0 +1,304 @@
+
+        /// setupterm(3) is a low-level call to begin doing any sort of `term.h`/`curses.h` work.
+        /// It's called internally by ncurses's `initscr()` and `newterm()`, but the C++ code called
+        /// it directly from [`initialize_curses_using_fallbacks()`].
+        pub fn setupterm(
+            term: *const libc::c_char,
+            filedes: libc::c_int,
+            errret: *mut libc::c_int,
+        ) -> libc::c_int;
+
+        /// Frees the `cur_term` TERMINAL  pointer.
+        pub fn del_curterm(term: *const core::ffi::c_void);
+
+        /// Checks for the presence of a termcap flag identified by the first two characters of
+        /// `id`. The C function returns an integer, but we just reinterpret that as a bool.
+        pub fn tgetflag(id: *const libc::c_char) -> bool;
+
+        /// Checks for the presence and value of a number capability in the termcap/termconf
+        /// database. A return value of `-1` indicates not found.
+        pub fn tgetnum(id: *const libc::c_char) -> libc::c_int;
+
+        pub fn tgetstr(
+            id: *const libc::c_char,
+            area: *mut *mut libc::c_char,
+        ) -> *const core::ffi::c_void;
+    }
+}
+
+// String capabilities
+pub const ENTER_ITALICS_MODE: StringCap = StringCap::new("ZH");
+pub const EXIT_ITALICS_MODE: StringCap = StringCap::new("ZR");
+pub const ENTER_DIM_MODE: StringCap = StringCap::new("mh");
+
+// Flag capabilities
+pub const EAT_NEWLINE_GLITCH: Flag = Flag::new("xn");
+
+pub struct Term {
+    overrides: Vec<(StringCap, String)>,
+}
+
+impl Term {
+    /// Internal constructor function. Like `Default` but only usable from within the module.
+    fn new() -> Self {
+        Term {
+            overrides: Vec::new(),
+        }
+    }
+
+    /// Looks up support for [`Capability`] `capability` in the termcap/terminfo database via the
+    /// curses library.
+    pub fn get<'a, C: Capability<'a>>(&'a mut self, capability: C) -> C::Result {
+        capability.lookup(self)
+    }
+
+    /// Overrides the string value of `capability` for the current terminal.
+    pub fn set(&mut self, id: StringCap, value: String) {
+        match self.overrides.binary_search_by(|entry| entry.0.cmp(&id)) {
+            Ok(idx) => self.overrides[idx] = (id, value),
+            Err(idx) => self.overrides.insert(idx, (id, value)),
+        }
+    }
+}
+
+enum Value<'a> {
+    String(&'a str),
+    Bool(bool),
+    Number(i32),
+}
+
+pub trait Capability<'a> {
+    type Result: Sized + 'a;
+    fn lookup(&self, term: &'a mut Term) -> Self::Result;
+}
+
+impl<'a> Capability<'a> for StringCap {
+    type Result = Option<&'a str>;
+
+    fn lookup(&self, term: &'a mut Term) -> Self::Result {
+        let id = self.0;
+        match term.overrides.binary_search_by(|entry| entry.0.cmp(self)) {
+            Ok(idx) => Some(&term.overrides[idx].1),
+            Err(idx) => {
+                let mut result = vec![b'\0'; 100];
```

Comment:
It's about a 10x conservative estimate based off my perusal of the `termcap` database. ncurses doesn't document the length and `tgetstr()` doesn't accept a length parameter to prevent an overflow!!

Reply:
I removed it from the latest update by not caching ncurses-returned strings at all.